posix/consolefs.c

```diff
@@ -0,0 +1,398 @@
+    if (!(buf = oe_malloc(buf_size)))
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Perform the read. */
+    if ((ret = _consolefs_read(desc, buf, buf_size)) <= 0)
+        goto done;
+
+    if (oe_iov_inflate(
+            buf, (size_t)ret, (struct oe_iovec*)iov, (size_t)iovcnt) != 0)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    if (buf)
+        oe_free(buf);
+
+    return ret;
+}
+
+static ssize_t _consolefs_writev(
+    oe_fd_t* desc,
+    const struct oe_iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = -1;
+    void* buf = NULL;
+    size_t buf_size = 0;
+
+    if (!iov || iovcnt < 0 || iovcnt > OE_IOV_MAX)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Create the write buffer from the IOV vector. */
+    if (oe_iov_deflate(iov, (size_t)iovcnt, &buf, &buf_size) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    ret = _consolefs_write(desc, buf, buf_size);
+
+done:
+
+    if (buf)
+        oe_free(buf);
+
+    return ret;
+}
+
+static oe_host_fd_t _consolefs_gethostfd(oe_fd_t* file_)
+{
+    oe_host_fd_t ret = -1;
+    file_t* file = _cast_file(file_);
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    ret = file->host_fd;
+
+done:
+    return ret;
+}
+
+static oe_off_t _consolefs_lseek(oe_fd_t* file_, oe_off_t offset, int whence)
+{
+    oe_off_t ret = -1;
+
+    OE_UNUSED(file_);
+    OE_UNUSED(offset);
+    OE_UNUSED(whence);
+    OE_RAISE_ERRNO(OE_ESPIPE);
+
+done:
+    return ret;
+}
+
+static int _consolefs_close(oe_fd_t* file_)
+{
+    int ret = -1;
+    file_t* file = _cast_file(file_);
+
+    if (!file)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Ask the host to perform this operation. */
+    {
+        if (oe_posix_close_ocall(&ret, file->host_fd) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        if (ret == -1)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Free the file structure. */
+    oe_free(file);
+
+done:
+    return ret;
+}
+
+static int _consolefs_getdents64(
+    oe_fd_t* file,
+    struct oe_dirent* dirp,
+    uint32_t count)
+{
+    OE_UNUSED(file);
+    OE_UNUSED(dirp);
+    OE_UNUSED(count);
+
+    /* The standard devices are not directories, so this is unsupported. */
+    OE_RAISE_ERRNO(OE_ENOTSUP);
+
+done:
+    return -1;
+}
+
+static oe_file_ops_t _ops = {
+    .fd.read = _consolefs_read,
+    .fd.write = _consolefs_write,
+    .fd.readv = _consolefs_readv,
+    .fd.writev = _consolefs_writev,
+    .fd.dup = _consolefs_dup,
+    .fd.ioctl = _consolefs_ioctl,
+    .fd.fcntl = _consolefs_fcntl,
+    .fd.close = _consolefs_close,
+    .fd.get_host_fd = _consolefs_gethostfd,
+    .lseek = _consolefs_lseek,
+    .getdents64 = _consolefs_getdents64,
+};
+
+static oe_file_ops_t _get_ops(void)
+{
+    return _ops;
+}
+
+static oe_fd_t* _new_file(uint32_t fileno)
+{
+    oe_fd_t* ret = NULL;
+    file_t* file = NULL;
+    struct _args
+    {
+        const char* path;
+        int flags;
+    };
+    struct _args args[] = {
+        {"/dev/stdin", OE_O_RDONLY},
```

Comment:
This is more of a general comment on the design of consolefs, since I'm getting confused due to conflicting things I'm seeing in the implementation:

1. Is the purpose of the consolefs to map to the host's stdin/stdout/stderr or for writing to a tty? 

- If it is the former, then you shouldn't error out on `_consolefs_lseek` and `_consolefs_getdents64`, because the host's stdin/stdout/stderr might be redirected to something that would actually cause those calls to succeed. For example, if you do `fstat(1, ...)` on a program that was executed like this `./my_program > my_file.txt`, you will see that stdout is a regular file and therefore seekable. Strangely enough, a program's standard fds can also be a directory! I think the best approach would be to have the syscall determine the appropriate error.

- If consolefs is for writing to a tty, then you should open `/dev/tty` for stdin/stdout/stderr. I'm not sure this is desirable however, because it breaks redirection and thus logging to a file. Also, if the process is running as a daemon, then there is no controlling tty, so I don't know what the expected behavior is there.

2. There are some issues with opening `/dev/[stdin/stdout/stderr]`, instead of duplicating `0,1,2`. They would have seperate entries in the kernel file description table, so they would have different file poisition, etc. This means that all the reads / writes to stdin/stdout/stderr are completely unsynchronized between the host and the guest. For example, if you have something like `./host ./enc > log.txt`, the enclave and host would overwrite the each other due to no file pointer synchronization. Further, `/dev/std...` are just symlinks to the process's 0,1,2 fds, so if the host closes or changes them before the enclave runs, the enclave would just pick up the changes. Personally, I think just doing `dup` on `0,1,2` would be a better alternative. There are some disadvantages of the dup approach though. For example, the enclave would pick up the fcntl settings and anything that changes the kernel file description object.

In my personal opinion, I think the simplest solution would be to duplicate file descriptions 0, 1, 2 to the enclave (ignore errors if you get bad fd). That would match the behavior we have today and I think that's what most people expect. You could also ocall add the ocalls to getdents and lseek for some minor edge cases. 

Eventually, we should have a discussion on how we want stdin/stdout/stderr to behave in the enclave.

Reply:
Thank you Akash. I created issue #1822 to track this beyond the scope of this PR. I also copied your comments to that issue.

> This is more of a general comment on the design of consolefs, since I'm getting confused due to conflicting things I'm seeing in the implementation:
> 
> Is the purpose of the consolefs to map to the host's stdin/stdout/stderr or for writing to a tty?
> 
This is to allow the enclave to write to the terminal device.
> 
> 
> If it is the former, then you shouldn't error out on _consolefs_lseek and _consolefs_getdents64, because the host's stdin/stdout/stderr might be redirected to something that would actually cause those calls to succeed. For example, if you do fstat(1, ...) on a program that was executed like this ./my_program > my_file.txt, you will see that stdout is a regular file and therefore seekable. Strangely enough, a program's standard fds can also be a directory! I think the best approach would be to have the syscall determine the appropriate error.
> 
> 
> If consolefs is for writing to a tty, then you should open /dev/tty for stdin/stdout/stderr. I'm not sure this is desirable however, because it breaks redirection and thus logging to a file. Also, if the process is running as a daemon, then there is no controlling tty, so I don't know what the expected behavior is there.
> 
> 
> 
> There are some issues with opening /dev/[stdin/stdout/stderr], instead of duplicating 0,1,2. They would have seperate entries in the kernel file description table, so they would have different file poisition, etc. This means that all the reads / writes to stdin/stdout/stderr are completely unsynchronized between the host and the guest. For example, if you have something like ./host ./enc > log.txt, the enclave and host would overwrite the each other due to no file pointer synchronization. Further, /dev/std... are just symlinks to the process's 0,1,2 fds, so if the host closes or changes them before the enclave runs, the enclave would just pick up the changes. Personally, I think just doing dup on 0,1,2 would be a better alternative. There are some disadvantages of the dup approach though. For example, the enclave would pick up the fcntl settings and anything that changes the kernel file description object.
> 
> In my personal opinion, I think the simplest solution would be to duplicate file descriptions 0, 1, 2 to the enclave (ignore errors if you get bad fd). That would match the behavior we have today and I think that's what most people expect. You could also ocall add the ocalls to getdents and lseek for some minor edge cases.
> Eventually, we should have a discussion on how we want stdin/stdout/stderr to behave in the enclave.

Reply:
I think the dup() solution might work too. But we probably should wait until the **hostfs** device is available. At that time, **consolefs** could leveragethe **hostfs** implementation. In the meantime, I think it is best not to reproduce the **hostfs** implementation (which handles getdents in portable way).

Reply:
One other thing: we intentionally avoid implementing **getdents** as an OCALL, since this would require a compatible implementation on Windows. We implement **getdents** within the enclave instead.